README.rst
Hiisi
======
The main idea behind hiisi module is to make the handling of hdf5 files as
fluent as possible. Module contains higher level tools such as search and
write methods that are build on top of h5py_ module.

HiisiHDF expands h5py.File class forms the base of the module.
Using HiisiHDF file handle, users can search metadata from the file
without any prior knowledge of the file structure. In addition, HiisiHDF contains
methods for listing the datasets and groups in the file and a convenient method
for creating and expanding hdf5 files.
 
HiisiHDF can be used as such or it can be used as base class for creating more
specialized file handles for different types of data files. An example of custom
data file handle is the hiisi.odim module that contains file handles for reading
weather radar data files that follow the OPERA odim data format.

.. _h5py: http://www.h5py.org/

Examples
--------
Open hdf5 file::

    >>> from hiisi import HiisiHDF
    >>> h5f = HiisiHDF('data.h5')

Get a list of datasets::

    >>> for dataset in h5f.datasets():
            print(dataset)
    '/dataset1/data1/data'
    '/dataset1/data2/data'
    '/dataset2/data1/data'
    '/dataset2/data2/data'

Retrieve all instances of an attribute::

    >>> gen = h5f.attr_gen('elangle')
    >>> pair = gen.next()
    >>> print(pair.path)
    '/dataset1/where'
    >>> print(pair.value)
    0.5

Create a new hdf5 file with content::

    >>> h5f = HiisiHDF('newfile.h5', 'w')
    >>> filedict = {'/':{'attr1':'A'},
                    '/dataset1/data1/data':{'DATASET':np.zeros(100), 'quantity':'emptyarray'}, 'B':'b'}
    >>> h5f.create_from_filedict(filedict)

Find the names of the groups that contain an attribute with a certain value::

    >>> for result in h5f.search('quantity', 'DBZH'):
            print(result)
    '/dataset1/data2/what'
    '/dataset2/data2/what'
    '/dataset3/data2/what'
    '/dataset4/data2/what'
    '/dataset5/data2/what'
        
Find the names of the groups that contain a numerical attribute
with a certain value or a value that is within the given tolerance::

    >>> h5f.search('elangle', 0.5, tolerance=0.1)
    [u'/dataset1/where']


Installation
------------
Hiisi is dependent on **numpy** and **h5py** packages. If you have these dependencies already 
installed you can simply run ``pip install hiisi``.

Installing the dependencies and the package

.. code-block:: bash

    $ install pip
    $ pip install numpy
    $ pip install Cython
    $ pip install h5py
    $ pip install hiisi

License
-------
This code is licensed under the MIT open source license.



changelog.rst
Changelog
=========

%%version%% (unreleased)
------------------------

- Added odim.rst. [Joonas Karjalainen]

- Removed hiisi/VERSION and MANIFEST.in. [Joonas Karjalainen]

- Changed the return type of search to iter. [Joonas Karjalainen]

- Made changes to the method of single sourcing version number. [Joonas
  Karjalainen]

- Renamed methods list_datasets to datasets and list_groups to groups
  and Changed their return type to iterator. [Joonas Karjalainen]

- Changed the indexing of OdimpPVOL.secotor and removed volume_slice
  method. [Joonas Karjalainen]

- Improved documentation. [Joonas Karjalainen]

- Improved documentation. [Joonas Karjalainen]

- Improved documentation. [Joonas Karjalainen]

- Improved documentation. [Joonas Karjalainen]

- Improved documentation. [Joonas Karjalainen]

- Improved documentation. [Joonas Karjalainen]

- Improved documentation. [Joonas Karjalainen]

- Improved documentation. [Joonas Karjalainen]

- Change in documentation. [Joonas Karjalainen]

- Made chages to odim.OdimPvol.volume_slice. [Joonas Karjalainen]

- Added methods to OdimPVOL. [Joonas Karjalainen]

- Refactored _set_elangles and select_dataset methods. [Joonas
  Karjalainen]

- Initial commit of odim module. [Joonas Karjalainen]

- Minor changes. [Joonas Karjalainen]

- Added a single source for package version number /hiisi/VERSION.
  [Joonas Karjalainen]

- Changed the topic of package to Scientific/Engineering. [Joonas
  Karjalainen]

- Documentation fixes. [Joonas Karjalainen]

- Added changelog. [Joonas Karjalainen]

- Fixed __version__ number from hiisi.py, added documentation. [Joonas
  Karjalainen]

- Fixed errors from doc strings. [Joonas Karjalainen]

- 'Multiple small fixes' [Joonas Karjalainen]

- Added tests. [Joonas Karjalainen]

- Added documentation and updated .gitignore. [Joonas Karjalainen]

- Added module files. [Joonas Karjalainen]

- Merge https://github.com/karjaljo/hiisi. [Joonas Karjalainen]

- Create .gitignore. [Joonas Karjalainen]

- Update LICENSE.md. [Joonas Karjalainen]

- Added README.rst. [Joonas Karjalainen]




tests/env.py
# -*- coding: utf-8 -*-
import sys
import os

# append module root directory to sys.path
sys.path.append(
    os.path.dirname(
        os.path.dirname(
            os.path.abspath(__file__)
        )
    )
)

tests/test_odim.py
# -*- coding: utf-8 -*-
import unittest
import env
import h5py
import hiisi
import numpy as np
import os

class Test(unittest.TestCase):
    
    def setUp(self):
        self.odim_file = hiisi.OdimPVOL('test_data/pvol.h5', 'r')
        
    def tearDown(self):
        self.odim_file.close()
        
    def test__set_elangles(self):
        self.odim_file._set_elangles()
        comparison_dict = {'A':self.odim_file['/dataset1/where'].attrs['elangle'],
                           'B':self.odim_file['/dataset2/where'].attrs['elangle'],
                            'C':self.odim_file['/dataset3/where'].attrs['elangle'],
                            'D':self.odim_file['/dataset4/where'].attrs['elangle'],
                            'E':self.odim_file['/dataset5/where'].attrs['elangle']
                            }
        self.assertDictEqual(self.odim_file.elangles, comparison_dict)
    
    def test__set_elangles_no_elangles(self):        
        with hiisi.OdimPVOL('empty_file.h5') as pvol:
            self.assertDictEqual(pvol.elangles, {})
            
    def test_select_dataset(self):
        self.assertEqual(self.odim_file.select_dataset('A', 'DBZH'), '/dataset1/data2/data')
        self.assertEqual(self.odim_file.select_dataset('B', 'VRAD'), '/dataset2/data3/data')
        self.assertEqual(self.odim_file.select_dataset('E', 'RHOHV'), '/dataset5/data7/data')
        
    def test_select_dataset_no_dataset_found(self):
        self.assertIsNone(self.odim_file.select_dataset('X', 'DBZH'))
        self.assertIsNone(self.odim_file.select_dataset('A', 'XXXX'))
 
    def test_sector_valid_indexes(self):
        filedict = {'/dataset1/data1/data':{'DATASET':np.arange(10*10).reshape((10,10))},
                    '/dataset1/where':{'elangle':0.5, 'rscale':500},
                    '/dataset1/data1/what':{'quantity':'DBZH'}
                    }        
        
        with hiisi.OdimPVOL('test_pvol.h5', 'w') as pvol:
            pvol.create_from_filedict(filedict)
            pvol._set_elangles()
            pvol.select_dataset('A', 'DBZH')
            # Single ray slice
            np.testing.assert_array_equal(pvol.sector(0, 0), np.arange(10).reshape(1, 10)) 
            # Multiple rays slice
            np.testing.assert_array_equal(pvol.sector(0, 1), np.arange(20).reshape((2, 10)))
            # Multiple rays, start distance given
            comparison_array = np.array([[4,  5,  6,  7,  8,  9],
                                         [14, 15, 16, 17, 18, 19]])
            np.testing.assert_array_equal(pvol.sector(0, 1, 2000), comparison_array)
            # Multiple rays, start and end distances given
            comparison_array = np.array([[4,  5,  6,  7],
                                         [14, 15, 16, 17]])
            np.testing.assert_array_equal(pvol.sector(0, 1, 2000, 4000), comparison_array)
            # Start slice bigger than end slice i.e. slice from 359->
            comparison_array = np.array([[80, 81, 82, 83, 84, 85, 86, 87, 88, 89],
                                         [90, 91, 92, 93, 94, 95, 96, 97, 98, 99],
                                            [ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
                                            [10, 11, 12, 13, 14, 15, 16, 17, 18, 19]])

            np.testing.assert_array_equal(pvol.sector(8, 1), comparison_array)           
                        
            
    def test_sector_invalid_indexes(self):
        filedict = {'/dataset1/data1/data':{'DATASET':np.arange(10*10).reshape((10,10))},
                    '/dataset1/where':{'elangle':0.5, 'rscale':500},
                    '/dataset1/data1/what':{'quantity':'DBZH'}
                    }        
        
        with hiisi.OdimPVOL('test_pvol.h5', 'w') as pvol:
            pvol.create_from_filedict(filedict)
            pvol._set_elangles()
            pvol.select_dataset('A', 'DBZH')
            
            # 
    def test_volume_slice(self):
        
        dataset1 = np.arange(10*10).reshape((10,10))
        dataset2 = np.arange(10*10).reshape((10,10))[:,:8] * 10
        dataset3 = np.arange(10*10).reshape((10,10))[:,:6] * 100
        
        layer1 = np.array([[ 0,  100,  200,  300,  400,  500],
                             [1000, 1100, 1200, 1300, 1400, 1500],
                            [2000, 2100, 2200, 2300, 2400, 2500],
                            [3000, 3100, 3200, 3300, 3400, 3500]])
        layer2 = np.array([[  0,  10,  20,  30,  40,  50],
                             [100, 110, 120, 130, 140, 150],
                            [200, 210, 220, 230, 240, 250],
                            [300, 310, 320, 330, 340, 350]]) 
        layer3 = np.array([[ 0,  1,  2,  3,  4,  5],
                             [10, 11, 12, 13, 14, 15],
                            [20, 21, 22, 23, 24, 25],
                            [30, 31, 32, 33, 34, 35]])
        comparison_array = np.dstack((layer1, layer2, layer3))
               
        filedict = {'/dataset1/data1/data':{'DATASET':dataset1},
                    '/dataset1/where':{'elangle':0.5, 'rscale':500},
                    '/dataset1/data1/what':{'quantity':'DBZH'},
                    '/dataset2/data1/data':{'DATASET':dataset2},
                    '/dataset2/where':{'elangle':0.7, 'rscale':500},
                    '/dataset2/data1/what':{'quantity':'DBZH'},
                    '/dataset3/data1/data':{'DATASET':dataset3},
                    '/dataset3/where':{'elangle':1.5, 'rscale':500},
                    '/dataset3/data1/what':{'quantity':'DBZH'},
                    }
        with hiisi.OdimPVOL('test_pvol.h5', 'w') as pvol:
            pvol.create_from_filedict(filedict)
            pvol._set_elangles()
            volume_slice = pvol.volume_slice('DBZH', 0, 3)
            np.testing.assert_array_equal(volume_slice, comparison_array)            
                                
        
        
if __name__=='__main__':
    unittest.main()

tests/test_hiisi.py
# -*- coding: utf-8 -*-
import unittest
import env
import hiisi
import h5py
import numpy as np
import uuid
import os


class Test(unittest.TestCase):
    
    def setUp(self):
        self.unique_attr_path = None
        self.unique_attr_value = None
        self.reoccuring_attr_paths = []
        self.reoccuring_attr_items = []
        self.dataset_paths = []
        self.group_paths = ['/']
        self.data_filename = 'hiisi_test_data.h5'

        self.create_hdf5_test_data()
        self.h5file = hiisi.hiisi.HiisiHDF(self.data_filename, 'r')
        print('run setUp')
        
        
    def create_hdf5_test_data(self):
        """Creates random hdf5 file for testing
        """
        n_branches = 3
        n_datasets = 3
        unique_attr = uuid.uuid1().hex
        reoccuring_attr = [uuid.uuid1().hex for x in range(n_branches)]
        dataset_data = np.zeros((3,3))
        h5f = h5py.File(self.data_filename, 'w')
        
        for i in range(n_branches):
            group_path = '/branch{}'.format(i)
            self.group_paths.append(group_path)
            branch = h5f.create_group(group_path)
            branch.attrs['reoccuring_attr'] = reoccuring_attr[i]
            self.reoccuring_attr_paths.append(branch.name)
            self.reoccuring_attr_items.append((branch.name, reoccuring_attr[i]))
            for j in range(n_datasets):
                dataset_name='/branch{}/data{}/dataset'.format(i, j)
                self.group_paths.append('/branch{}/data{}'.format(i, j))
                dataset = h5f.create_dataset(dataset_name, data=np.int8(dataset_data), dtype='int8')
                self.dataset_paths.append(dataset.name)                    
                if i==1 and j==1:
                    dataset.attrs['unique_attr'] = unique_attr
                    self.unique_attr_path = dataset.name
                    self.unique_attr_value = unique_attr
        
        h5f.close()
        
    def tearDown(self):
        os.remove(self.data_filename)
                         
    def test_is_unique_attribute_true(self):
        self.assertTrue(self.h5file.is_unique_attr('unique_attr'))
        
    def test_is_unique_attribute_false(self):
        self.assertFalse(self.h5file.is_unique_attr('reoccuring_attr'))
        self.assertFalse(self.h5file.is_unique_attr('not_existing_attr'))
        
    def test_attr_exists_true(self):
        self.assertTrue(self.h5file.attr_exists('unique_attr'))        
            
    def test_attr_exists_false(self):
        self.assertFalse(self.h5file.attr_exists('not_existing_attr'))    

    def test_datasets(self):
        self.assertItemsEqual(list(self.h5file.datasets()), self.dataset_paths)
        
    def test_datasets_no_datasets_found(self):
        with hiisi.HiisiHDF('tmp.h5', 'w') as h5f:
            self.assertItemsEqual(list(h5f.datasets()), [])
        os.remove('tmp.h5')
        
    def test_groups(self):
        self.assertItemsEqual(list(self.h5file.groups()), self.group_paths)
        
    def test_groups_no_groups_found(self):
        with hiisi.HiisiHDF('tmp.h5', 'w') as h5f:
            self.assertItemsEqual(h5f.groups(), ['/'])
        os.remove('tmp.h5')

    def test_attr_gen(self):
        attr_gen = self.h5file.attr_gen('reoccuring_attr')
        attr_items = []     
        for i in attr_gen:
            attr_items.append((i.path, i.value))
        self.assertListEqual(attr_items, self.reoccuring_attr_items)

    def test_attr_gen_no_match(self):
        attr_gen = self.h5file.attr_gen('not_existing_attr')        
        with self.assertRaises(StopIteration):
            attr_gen.next()
            
    def test_create_from_filedict_new_file(self):
        filename = 'create_from_filedict_test.h5'
        with hiisi.HiisiHDF(filename, 'w') as h5f:
            file_dict = {}
            file_dict['/'] = {'A':1, 'B':2}
            file_dict['/dataset1/data1/data'] = {'DATASET':np.arange(9).reshape((3,3)), 'C':'c'}
            file_dict['/dataset1/data1/what'] = {'D':123}
            h5f.create_from_filedict(file_dict)        

        with hiisi.HiisiHDF(filename, 'r') as h5f:      
            self.assertEqual(h5f['/'].attrs['A'], 1)
            self.assertEqual(h5f['/'].attrs['B'], 2)
            self.assertEqual(h5f['/dataset1/data1/data'].attrs['C'], 'c')
            np.testing.assert_array_equal(h5f['/dataset1/data1/data'][:], np.arange(9).reshape((3,3)))
            self.assertEqual(h5f['/dataset1/data1/what'].attrs['D'], 123)
        os.remove(filename)

    
    def test_create_from_filedict_append_new_goup(self):
        filename = 'create_from_filedict_test.h5'
        # Create the file
        with hiisi.HiisiHDF(filename, 'w') as h5f:
            file_dict = {}
            file_dict['/'] = {'A':1, 'B':2}
            file_dict['/dataset1/data1/data'] = {'DATASET':np.arange(9).reshape((3,3)), 'C':'c'}
            file_dict['/dataset1/data1/what'] = {'D':123}
            h5f.create_from_filedict(file_dict)        

        # Append the file created above
        with hiisi.HiisiHDF(filename) as h5f:
            h5f.create_from_filedict({'/added_group':{'attr1':1}})
            
        # Check the results
        with hiisi.HiisiHDF(filename, 'r') as h5f:      
            self.assertEqual(h5f['/'].attrs['A'], 1)
            self.assertEqual(h5f['/'].attrs['B'], 2)
            self.assertEqual(h5f['/dataset1/data1/data'].attrs['C'], 'c')
            np.testing.assert_array_equal(h5f['/dataset1/data1/data'][:], np.arange(9).reshape((3,3)))
            self.assertEqual(h5f['/dataset1/data1/what'].attrs['D'], 123)
            self.assertEqual(h5f['/added_group'].attrs['attr1'], 1)
            
        os.remove(filename)
        
    def test_create_from_filedict_modify_existing_content(self):
        filename = 'create_from_filedict_test.h5'
        # Create the file
        with hiisi.HiisiHDF(filename, 'w') as h5f:
            file_dict = {}
            file_dict['/'] = {'A':1, 'B':2}
            file_dict['/dataset1/data1/data'] = {'DATASET':np.arange(9).reshape((3,3)), 'C':'c'}
            file_dict['/dataset1/data1/what'] = {'D':123}
            h5f.create_from_filedict(file_dict)        

        # Append the file created above
        with hiisi.HiisiHDF(filename) as h5f:
            h5f.create_from_filedict({'/dataset1/data1/data':{'C':'new_value'}})
            
        # Check the results
        with hiisi.HiisiHDF(filename, 'r') as h5f:      
            self.assertEqual(h5f['/'].attrs['A'], 1)
            self.assertEqual(h5f['/'].attrs['B'], 2)
            self.assertEqual(h5f['/dataset1/data1/data'].attrs['C'], 'new_value')
            np.testing.assert_array_equal(h5f['/dataset1/data1/data'][:], np.arange(9).reshape((3,3)))
            self.assertEqual(h5f['/dataset1/data1/what'].attrs['D'], 123)
            
        os.remove(filename)
           
    def test_search_no_match(self):
        self.assertListEqual([], list(self.h5file.search('madeupkey', 'xyz')))
            
    def test_search_single_match(self):
        self.assertListEqual([self.unique_attr_path], list(self.h5file.search('unique_attr', self.unique_attr_value)))
            
    def test_search_multiple_matches(self):
        filename = 'test_search_multiple_matches.h5'
        with hiisi.HiisiHDF(filename) as h5f:
            groups = ['/group1', '/group2', '/basegroup/group1']
            for g in groups:
                group = h5f.create_group(g)
                group.attrs['attribute']  = 'attribute'
            
        with hiisi.HiisiHDF(filename) as h5f:
            self.assertListEqual(sorted(groups), sorted(list(h5f.search('attribute', 'attribute'))))
        os.remove(filename)
            
    def test_search_numerical_attribute_within_tolerance(self):
        filename = 'test_search_numerical_attribute.h5'
        with hiisi.HiisiHDF(filename) as h5f:
            group = h5f.create_group('/group1')
            group.attrs['attribute']  = 7.3
            group = h5f.create_group('/group2')
            group.attrs['attribute']  = 0.1001
            group = h5f.create_group('/basegroup/group1')        
            group.attrs['attribute']  = 0.5245

        with hiisi.HiisiHDF(filename) as h5f:        
            self.assertListEqual(['/basegroup/group1'], list(h5f.search('attribute', 0.5, 0.1)))
        os.remove(filename)

    def test_search_numerical_attribute_outside_tolerance(self):            
        filename = 'test_search_numerical_attribute.h5'
        with hiisi.HiisiHDF(filename) as h5f:
            group = h5f.create_group('/group1')
            group.attrs['attribute']  = 7.3
            group = h5f.create_group('/group2')
            group.attrs['attribute']  = 0.1001
            group = h5f.create_group('/basegroup/group1')        
            group.attrs['attribute']  = 0.5245

        with hiisi.HiisiHDF(filename) as h5f:        
            self.assertListEqual([], list(h5f.search('attribute', 7, 0.1)))
        os.remove(filename)
        
    
if __name__=='__main__':
    unittest.main()       

hiisi/hiisi.py
# -*- coding: utf-8 -*-
import h5py
import os
from collections import namedtuple
PathValue = namedtuple('PathValue', ['path', 'value'])


class HiisiHDF(h5py.File):
    """hdf5 file handle written on top of h5py.File.

    Module offers easy to use search, and write methods for handling
    HDF5 files.
    """
    CACHE = {'search_attribute':None,
             'dataset_paths':[],
             'group_paths':[],
             'attribute_paths':[]}

    def __init__(self, *args, **kwargs):
        super(HiisiHDF, self).__init__(*args, **kwargs)

    @staticmethod
    def _clear_cache():
        HiisiHDF.CACHE = {'search_attribute':None,
                          'dataset_paths':[],
                          'group_paths':[],
                          'attribute_paths':[]}

    @staticmethod
    def _is_dataset(name, obj):
        if isinstance(obj, h5py.Dataset):
            HiisiHDF.CACHE['dataset_paths'].append(obj.name)

    @staticmethod
    def _is_group(name, obj):
        if isinstance(obj, h5py.Group):
            HiisiHDF.CACHE['group_paths'].append(obj.name)

    @staticmethod
    def _find_attr_paths(name, obj):
        if HiisiHDF.CACHE['search_attribute'] in obj.attrs:
            HiisiHDF.CACHE['attribute_paths'].append(obj.name)

    @staticmethod
    def _is_attr_path(name, obj):
        if HiisiHDF.CACHE['search_attribute'] in obj.attrs:
            return obj.name

    def attr_exists(self, attr):
        """Returns True if at least on instance of the attribute is found
        """
        gen = self.attr_gen(attr)
        n_instances = len(list(gen))
        if n_instances > 0:
            return True
        else:
            return False

    def is_unique_attr(self, attr):
        """Returns true if only single instance of the attribute is found
        """
        gen = self.attr_gen(attr)
        n_instances = len(list(gen))
        if n_instances == 1:
            return True
        else:
            return False

    def datasets(self):
        """Method returns an iterator over all dataset paths.

        Examples
        --------
        >>> for dataset in h5f.datasets():
                print(dataset)
        '/dataset1/data1/data'
        '/dataset1/data2/data'
        '/dataset2/data1/data'
        '/dataset2/data2/data'
        """
        HiisiHDF._clear_cache()
        self.visititems(HiisiHDF._is_dataset)
        return iter(HiisiHDF.CACHE['dataset_paths'])

    def groups(self):
        """Method returns an iterator over all goup paths
        
        Examples
        --------        
        >>> for group in h5f.groups():
                print(group)        
        '/'
        '/dataset1'
        '/dataset1/data1'
        '/dataset1/data2'
        """
        HiisiHDF._clear_cache()
        self.CACHE['group_paths'].append('/') #Every hdf5 file has a root
        self.visititems(HiisiHDF._is_group)
        return iter(HiisiHDF.CACHE['group_paths'])


    def attr_gen(self, attr):
        """Returns attribute generator that yields namedtuples containing
        path value pairs
        
        Parameters
        ----------
        attr : str
            Name of the search attribute

        Returns
        -------
        attr_generator : generator
            Returns a generator that yields named tuples with field names
            path and value.

        Examples
        --------
        >>> gen = h5f.attr_gen('elangle')
        >>> pair = next(gen)
        >>> print(pair.path)
        '/dataset1/where'
        >>> print(pair.value)
        0.5

        """
        HiisiHDF._clear_cache()
        HiisiHDF.CACHE['search_attribute'] = attr
        HiisiHDF._find_attr_paths('/', self['/']) # Check root attributes
        self.visititems(HiisiHDF._find_attr_paths)
        path_attr_gen = (PathValue(attr_path, self[attr_path].attrs.get(attr)) for attr_path in HiisiHDF.CACHE['attribute_paths'])
        return path_attr_gen


    def create_from_filedict(self, filedict):
        """
        Creates h5 file from dictionary containing the file structure.
        
        Filedict is a regular dictinary whose keys are hdf5 paths and whose
        values are dictinaries containing the metadata and datasets. Metadata
        is given as normal key-value -pairs and dataset arrays are given using
        'DATASET' key. Datasets must be numpy arrays.
                
        Method can also be used to append existing hdf5 file. If the file is
        opened in read only mode, method does nothing.

        Examples
        --------
        Create newfile.h5 and fill it with data and metadata

        >>> h5f = HiisiHDF('newfile.h5', 'w')
        >>> filedict = {'/':{'attr1':'A'},
                        '/dataset1/data1/data':{'DATASET':np.zeros(100), 'quantity':'emptyarray'}, 'B':'b'}
        >>> h5f.create_from_filedict(filedict)

        """
        if self.mode in ['r+','w', 'w-', 'x', 'a']:
            for h5path, path_content in filedict.iteritems():
                if path_content.has_key('DATASET'):
                    # If path exist, write only metadata
                    if h5path in self:
                        for key, value in path_content.iteritems():
                            if key != 'DATASET':
                                self[h5path].attrs[key] = value
                    else:
                        try:
                            group = self.create_group(os.path.dirname(h5path))
                        except ValueError:
                            group = self[os.path.dirname(h5path)]
                            pass # This pass has no effect?
                        new_dataset = group.create_dataset(os.path.basename(h5path), data=path_content['DATASET'])
                        for key, value in path_content.iteritems():
                            if key != 'DATASET':
                                new_dataset.attrs[key] = value
                else:
                    try:  
                        group = self.create_group(h5path)
                    except ValueError:
                        group = self[h5path]
                    for key, value in path_content.iteritems():
                        group.attrs[key] = value

    def search(self, attr, value, tolerance=0):
        """Find paths with a key value match

        Parameters
        ----------
        attr : str
            name of the attribute
        value : str or numerical value
            value of the searched attribute
        
        Keywords
        --------
        tolerance : float
            tolerance used when searching for matching numerical
            attributes. If the value of the attribute found from the file
            differs from the searched value less than the tolerance, attributes
            are considered to be the same.

        Returns
        -------
        result_iter : iterator
            An iterator over matching paths

        Examples
        --------

        >>> for result in h5f.search('elangle', 0.5, 0.1):
                print(result)        
        '/dataset1/where'

        >>> for result in h5f.search('quantity', 'DBZH'):
                print(result)
        '/dataset1/data2/what'
        '/dataset2/data2/what'
        '/dataset3/data2/what'
        '/dataset4/data2/what'
        '/dataset5/data2/what'
        
        """
        found_paths = []
        gen = self.attr_gen(attr)
        for path_attr_pair in gen:
            # if attribute is numerical use numerical_value_tolerance in
            # value comparison. If attribute is string require exact match
            if isinstance(path_attr_pair.value, str):
                type_name = 'str'
            else:
                type_name = path_attr_pair.value.dtype.name
            if 'int' in type_name or 'float' in type_name:
                if abs(path_attr_pair.value - value) <= tolerance:
                    found_paths.append(path_attr_pair.path)
            else:
                if path_attr_pair.value == value:
                    found_paths.append(path_attr_pair.path)

        return iter(found_paths)

